File: YJ/20200904/Graph/순위.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

int solution(int n, vector<vector<int>> results) {
    int answer = 0;
    vector<vector<bool>> can_go(n + 1, vector<bool>(n + 1, false));
    
    for (int i = 0; i < results.size(); i ++) {
        can_go[results[i][0]][results[i][1]] = true;
    }
    
    for (int k = 1; k <= n; k ++) {
        for (int i = 1; i <= n; i ++) {
            for (int j = 1; j <= n; j ++) {
                if (can_go[i][k] && can_go[k][j])
                    can_go[i][j] = true;
            }
        }
    }
    
    for (int i = 1; i <= n; i ++) {
        int cnt = 0;
        for (int j = 1; j <= n; j ++) {
            if (can_go[i][j] || can_go[j][i]) cnt ++;
        }
        if (cnt == n - 1) answer ++;
    }
    return answer;
}

File: JY/8/8.28/순위.py
from collections import defaultdict

def solution(n, results):
    ans = 0
    win = defaultdict(list)
    lose = defaultdict(list)
    for i, j in results:
        win[i].append(j)
        lose[j].append(i)
        
    for i in range(1, n+1):
        stack = []
        stack.append(i)
        visit = [False]*(n+1)
        cnt = 0
        while stack:
            node = stack.pop()
            for j in win[node]:
                if not visit[j]:
                    cnt += 1
                    visit[j] = True
                    stack.append(j)
        stack = []
        stack.append(i)
        visit = [False]*(n+1)
        while stack:
            node = stack.pop()
            for j in lose[node]:
                if not visit[j]:
                    cnt += 1
                    visit[j] = True
                    stack.append(j)
        if cnt == (n-1):
            ans += 1
        
    
    return ans